curvlinops/diagonal/hutchinson.py

```diff
@@ -33,54 +36,46 @@ class HutchinsonDiagonalEstimator:
+
     Example:
-        >>> from numpy import diag, mean, round
+        >>> from numpy import diag
         >>> from numpy.random import rand, seed
         >>> from numpy.linalg import norm
         >>> seed(0) # make deterministic
-        >>> A = rand(10, 10)
+        >>> A = rand(40, 40)
         >>> diag_A = diag(A) # exact diagonal as reference
-        >>> estimator = HutchinsonDiagonalEstimator(A)
         >>> # one- and multi-sample approximations
-        >>> diag_A_low_precision = estimator.sample()
-        >>> samples = [estimator.sample() for _ in range(1_000)]
-        >>> diag_A_high_precision = mean(samples, axis=0)
+        >>> diag_A_low_precision = hutchinson_diag(A, num_matvecs=1)
+        >>> diag_A_high_precision = hutchinson_diag(A, num_matvecs=30)
         >>> # compute residual norms
-        >>> error_low_precision = norm(diag_A - diag_A_low_precision)
-        >>> error_high_precision = norm(diag_A - diag_A_high_precision)
+        >>> error_low_precision = norm(diag_A - diag_A_low_precision) / norm(diag_A)
+        >>> error_high_precision = norm(diag_A - diag_A_high_precision) / norm(diag_A)
         >>> assert error_low_precision > error_high_precision
         >>> round(error_low_precision, 4), round(error_high_precision, 4)
-        (5.7268, 0.1525)
+        (4.616, 1.2441)
     """
-
-    def __init__(self, A: LinearOperator):
-        """Store the linear operator whose diagonal will be estimated.
-
-        Args:
-            A: Linear square-shaped operator whose diagonal will be estimated.
-
-        Raises:
-            ValueError: If the operator is not square.
-        """
-        if len(A.shape) != 2 or A.shape[0] != A.shape[1]:
-            raise ValueError(f"A must be square. Got shape {A.shape}.")
-        self._A = A
-
-    def sample(self, distribution: str = "rademacher") -> ndarray:
-        """Draw a sample from the diagonal estimator.
-
-        Multiple samples can be combined into a more accurate diagonal estimation via
-        averaging.
-
-        Args:
-            distribution: Distribution of the vector along which the linear operator
-                will be evaluated. Either ``'rademacher'`` or ``'normal'``.
-                Default is ``'rademacher'``.
-
-        Returns:
-            A Sample from the diagonal estimator.
-        """
-        dim = self._A.shape[1]
-        v = random_vector(dim, distribution)
-        Av = self._A @ v
-        return v * Av
+    if len(A.shape) != 2 or A.shape[0] != A.shape[1]:
+        raise ValueError(f"A must be square. Got shape {A.shape}.")
+    dim = A.shape[1]
+    if num_matvecs >= dim:
+        raise ValueError(
+            f"num_matvecs ({num_matvecs}) must be less than A's size ({dim})."
+        )
+    G = column_stack([random_vector(dim, distribution) for _ in range(num_matvecs)])
+    AG = A @ G
+    return einsum("ij,ij->i", G, AG) / num_matvecs
```

Comment:
Can be rewritten as a single einsum.

Reply:
You are right if `A` is a `Tensor`. However, in general `A` can be a `LinearOperator`. In this case, we must compute `A @ G`, because expanding `A`'s matrix representation would be too costly.

(The same applies to your other comments on single einsums, I believe. Let me know if I missed something here.)